### DISCUSSION TITLE
Return Values from Popups?

### DISCUSSION THREAD

**Discussion by LuckyDucko** [LuckyDucko/Mopups#8] [Polls]
I feel that with Popups, many people want them to be able to return a value to a page, with an example being a popup with a Yes/No question, and the user response is sent to the previous page.

The communityToolkit popup has these, and I also added this ability in the AwaitablePopups plugin.

I would prefer to use TaskCompletionSource over anything else, as the thought of using the MessagingCentre is just asking for a bad time. 

I am eager to hear anyones thoughts, especially @AswinPG & @maxkoshevoi

**Comment by LuckyDucko** [comment]
Update

I am creating AwaitablePopups for Mopups, which will be named Mopups.Awaitable.

It will operate very similarly to the original repo and such, as I dont intend on bringing any huge changes to it. That repo is designed for customisable awaitable popups based on Viewmodels being attached to specific popups pages. Its main strength will be that it has some predesigned use cases that can be picked apart, extended or used as is. 

If we add tasks to this repo, it will be a very barebones version in comparison.

**Comment by IeuanWalker** [comment]
I'm already achieving this by passing in an event, but would be nice to just await the returned value.


https://user-images.githubusercontent.com/6544051/178718103-449e21ae-e5a0-4e37-a08a-e8e6f4b99577.mp4



Creating the popup - 

```csharp
PickerPopup pickerPopup = new("Select a theme:", new List<PickerList>
{
	new()
	{
		IsSelected = _viewModel.Theme.Equals(PreferredTheme.Default),
		ReturnValue = PreferredTheme.Default,
		DisplayText = nameof(PreferredTheme.Default)
	},
	new()
	{
		IsSelected = _viewModel.Theme.Equals(PreferredTheme.Light),
		ReturnValue = PreferredTheme.Light,
		DisplayText = nameof(PreferredTheme.Light)
	},
	new()
	{
		IsSelected = _viewModel.Theme.Equals(PreferredTheme.Dark),
		ReturnValue = PreferredTheme.Dark,
		DisplayText = nameof(PreferredTheme.Dark)
	}
},
(object? sender, object? e) =>
{
	if (e is not null)
	{
		_viewModel.Theme = (PreferredTheme)e;
	}
});

await MopupService.Instance.PushAsync(pickerPopup);
```

Selecting an item in the popup - 
```csharp
async void ItemSelected_Tapped(object sender, EventArgs e)
{
	var grid = (Grid)sender;
	var context = (PickerList)grid.BindingContext;

	SelectedItemEvent.Invoke(this, context.ReturnValue);

	await MopupService.Instance.PopAsync();
}
```

**Reply by LuckyDucko** [reply]
I'm thinking of perhaps expanding the api to have 

`await MopupService.Instance.PushAwaitableAsync` or `await MopupService.Instance.PushReturnableAsync`

Need to figure out what the best language is there for it.

**Reply by rupisaini123** [reply]
@IeuanWalker I am looking for same animation that you have added in the sample. Can you please share the sample code for this animation, when popup opens and closes? 
Thanks in advance.

**Reply by IeuanWalker** [reply]
@rupisaini123 this should work -
```xaml
    <Mopups:PopupPage.Animation>
        <Animations:MoveAnimation DurationIn="400"
                                  DurationOut="300"
                                  EasingIn="SinOut"
                                  EasingOut="SinIn"
                                  HasBackgroundAnimation="True"
                                  PositionIn="Bottom"
                                  PositionOut="Bottom" />
    </Mopups:PopupPage.Animation>
```

**Reply by tanakamawere** [reply]
Hie there @IeuanWalker 
Mopups have bottom sheets? Please share how exactly you implemented them

And if you have a link to a more in-depth wiki, I would appreciate it

**Reply by IeuanWalker** [reply]
@tanakamawere not directly no.

You can use the code above to return a value back from the popup.

And you can use the code on this issue to set the height of the bottom sheet - https://github.com/LuckyDucko/Mopups/issues/12
All you need to do is add a collectionview to the popup and use the event approach above to return the select value.

**Comment by enisn** [comment]
I have a simple implementation with TaskCompletionSource:
https://github.com/enisn/UraniumUI/blob/4269167f471f200cd47b7b9341a300b8bba10e19/src/UraniumUI.Dialogs.Mopups/MopupsDialogExtensions.cs#L16-L50

**Comment by LuckyDucko** [comment]
I added my Mopups.prebaked addition into this plugin

It is a bit overkill...
 and does require people making things a certain way which could be off putting....
 and probably could do with some work to make it easier to integrate with other peoples work...
 
 However, for most use cases it provides a prebuilt way to do login, acknowledgements, inputs, even a basic login. It also provides ways to overload that and use the underlying code with your own xaml files.
